lib/layout/forceDirectedLayout.js

```diff
@@ -0,0 +1,175 @@
+//layout/SpringLayout.java?r=19
+
+function layOutGraph(graph){
+  mGraph = graph;
+  vertices = graph.get('vertices');
+  var edgeList = graph.get('edges');
+  var runner = true;
+  while(runner){
+    change = false;
+    for(var i = 0; i < edgeList.length; i++){
+      springForce( edgeList[i] );
+    }
+    if(change === false){
+      runner = false;
+    }
+  }
+  //defaulting all ports that do not have a position
+  for(var j = 0; j < vertices.length; j++){
+    var inputs = vertices[j].get('inputs');
+    var outputs = vertices[j].get('outputs');
+
+    for(var h = 0; h < inputs.length; h++){
+      if(inputs[h].get('position') === null){
+        inputs[h].set('position', new CardinalPortPosition('E', 50));
+      }
+    }
+
+    for(h = 0; h < outputs.length; h++){
+      if(outputs[h].get('position') === null){
+        outputs[h].set('position', new CardinalPortPosition('W', 50));
+      }
+    }
+  }
+}
+
+function springForce(edge){
+  //vertex to
+  var vertexTo = mGraph.getVertexById(edge.get('to').get('ownerVertexId'));
+  //vertex from
+  var vertexFrom = mGraph.getVertexById(edge.get('from').get('ownerVertexId'));
+  //x distance between them
+  var xLength = edgeXLength(edge);
+  //y distance between them
+  var yLength = edgeYLength(edge);
+  //straight line edge length
+  var length = edgeLength(edge);
+  //checks for both pinned
+  if(vertexTo.get('isPinned') === true && vertexFrom.get('isPinned') === true){
+    return;
+  }
+
+  //hacky solution for length being 0
+  if(length === 0){
+    length = 0.99;
+  }
+  //repulsion force calculation
+  var force = forceStr * (iLength - length) / length;
+  var pwrTo = vertexTo.get('inputs').length + vertexTo.get('outputs').length;
+  var pwrFr = vertexFrom.get('inputs').length + vertexFrom.get('outputs').length;
+  force = force * Math.pow(stretchForce, (pwrTo + pwrFr - 4));
+
+  //repulsion change in x
+  var repulsionX = force * xLength;
+  //repulsion change in y
+  var repulsionY = force * yLength;
+  //attraction x
+  var attractionX;
+  //attraction y
+  var attractionY;
+  //math for min length
+  if(length < 1){
+    attractionX = Math.random() * 10 + 1;
+    attractionY = Math.random() * 10 + 1;
+  } else {
+    attractionX = xLength / (length * length);
+    attractionY = yLength / (length * length);
+  }
+  var len = attractionX * attractionX + attractionY * attractionY;
+  //ensures length is positive
+  if(len > 0){
+    len = Math.sqrt(len);
+    attractionX = attractionX / len;
+    attractionY = attractionY / len;
+  }
+  //repulsion minus attraction
+  repulsionX = repulsionX - attractionX;
+  repulsionY = repulsionY - attractionY;
+  var lim = 0.1;
+
+  if(vertexTo.get('isPinned') === false && vertexFrom.get('isPinned') === false){
+    vertexTo.set('position', {
+                              'x': (vertexTo.get('position').x + repulsionX),
+                              'y': (vertexTo.get('position').y + repulsionY)
```

Comment:
Getting atttribute multiple times. Recommend storing in variable. This will be cleaner and faster, as the complexity of the lookup function may be high (it's not *that* high right now but you should not assume that will remain true).